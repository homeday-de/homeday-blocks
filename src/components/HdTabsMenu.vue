<template>
  <div class="tabs-menu">
    <div class="tabs-menu__wrap">
      <div class="tabs-menu__wrap__inner">
        <flickity
          ref="flickity"
          :options="flickityOptions"
        >
          <button
            v-for="(item, i) in items"
            :key="i"
            :class="{
              'tabs-menu__item': true,
              'btn': true,
              'btn--flat': true,
              'js-tabs-menu-item': true,
              'isActive': item.value === value
            }"
            :data-index="i"
            type="button"
            @click.prevent="$emit('input', item.value)"
          >
            {{ item.label }}
          </button>
        </flickity>
      </div>
    </div>
  </div>
</template>

<script>
import { hidePaginationWhenNotNeeded } from 'homeday-blocks/src/services/flickity';
// eslint-disable-next-line import/extensions
import Flickity from 'vue-flickity';

const TABS_MENU_ITEM_CLASS = 'js-tabs-menu-item';

export default {
  name: 'HdTabsMenu',
  components: {
    Flickity,
  },
  props: {
    items: {
      type: Array,
      default: () => [],
    },
    value: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      flickityOptions: {
        prevNextButtons: true,
        pageDots: false,
        cellAlign: 'center',
        freeScroll: true,
        contain: true,
        initialIndex: this.items.findIndex(item => item.value === this.value),
        on: {
          ready: hidePaginationWhenNotNeeded,
          resize: hidePaginationWhenNotNeeded,
        },
      },
    };
  },
  watch: {
    value(value) {
      this.$refs.flickity.select(this.items.findIndex(item => item.value === value));
    },
  },
  mounted() {
    this.addFocusEvents();
  },
  beforeDestroy() {
    this.removeFocusEvents();
  },
  methods: {
    bringCellIntoViewOnFocus(event) {
      if (!event.target.classList.contains(TABS_MENU_ITEM_CLASS)) {
        return;
      }

      this.$refs.flickity.select(parseInt(event.target.getAttribute('data-index'), 10));
    },
    addFocusEvents() {
      this.$el.addEventListener(
        'focus',
        this.bringCellIntoViewOnFocus,
        true,
      );
    },
    removeFocusEvents() {
      this.$el.removeEventListener(
        'focus',
        this.bringCellIntoViewOnFocus,
        true,
      );
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.tabs-menu {
  $_root: &;
  position: relative;

  &__item {
    padding-left: 0;
    padding-right: 0;
    margin-right: $inline-l;
    white-space: nowrap;
    font-weight: 400;
    text-align: left;
    outline: 0;

    &::after {
      width: 100%;
    }

    &:focus::after,
    &.isActive::after {
      border-color: getShade($secondary-color, 110);
    }
  }

  .flickity--no-controls .flickity-slider {
    transform: none !important;
  }

  svg.flickity-button-icon {
    display: none;
  }

  .flickity-prev-next-button {
    height: 100%;
    padding: 0;
    background-color: transparent;

    &:active {
      opacity: 1;
    }

    &:focus {
      border-radius: 0;
      box-shadow: 0 4px 0 0 getShade($secondary-color, 110);
      outline: initial;
    }

    &:disabled {
      opacity: 0;
      pointer-events: none;
    }

    &::before {
      display: block;
      position: absolute;
      top: 0;
      right: 0;
      width: 24px;
      height: 100%;
      background-color: $primary-bg;
      background-image: url('~homeday-blocks/src/assets/small-icons/ic_chevron.svg');
      background-repeat: no-repeat;
      background-position: center left;
      background-size: contain;
      content: '';
    }

    &::after {
      display: block;
      position: absolute;
      top: 0;
      right: 24px;
      width: 16px;
      height: 100%;
      background-image: linear-gradient(
        -90deg,
        $primary-bg,
        rgba($primary-bg, 0)
      );
      content: '';
    }

    &.previous {
      left: 0;

      &::before {
        transform: rotate(180deg);
        right: auto;
        left: 0;
      }

      &::after {
        left: 24px;
        right: auto;
        transform: rotate(180deg);
      }
    }

    &.next {
      right: 0;
    }
  }

  .flickity--no-controls .flickity-prev-next-button {
    display: none;
  }
}
</style>
