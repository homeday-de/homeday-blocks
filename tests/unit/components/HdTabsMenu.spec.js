import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import HdTabsMenu from '@/components/HdTabsMenu.vue';
import ITEMS from '@/stories/mocks/MENU_ITEMS';

const MENU_ITEM_SELECTOR = '.js-tabs-menu-item';
const ACTIVE_MENU_ITEM_SELECTOR = '.js-tabs-menu-item.isActive';

const wrapperBuilder = wrapperFactoryBuilder(HdTabsMenu, {
  props: {
    items: ITEMS,
  },
});

describe('HdTabsMenu', () => {
  it('The component is rendered', () => {
    const wrapper = wrapperBuilder({
      props: {
        value: ITEMS[1].value,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Rendered selected menu item should match passed in preselected item', () => {
    const wrapper = wrapperBuilder({
      props: {
        value: ITEMS[1].value,
      },
    });

    expect(wrapper.findAll(ACTIVE_MENU_ITEM_SELECTOR).length).toBe(1);
  });

  it('By default, no menu item is selected', () => {
    const wrapper = wrapperBuilder();

    expect(wrapper.findAll(ACTIVE_MENU_ITEM_SELECTOR).length).toBe(0);
  });

  it('Menu item can be selected', async () => {
    const wrapper = wrapperBuilder();

    // select a menu item
    wrapper.find(MENU_ITEM_SELECTOR).trigger('click');
    // confirm that input event is emitted
    expect(wrapper.emitted('input')).toBeTruthy();
    // And that expected payload is passed with the event
    const payload = wrapper.emitted('input')[0][0];
    expect(payload).toBe(ITEMS[0].value);
    // check that a menu item is selected
    wrapper.setProps({ value: payload });

    await wrapper.vm.$nextTick();

    expect(wrapper.findAll(ACTIVE_MENU_ITEM_SELECTOR).length).toBe(1);
  });
});
