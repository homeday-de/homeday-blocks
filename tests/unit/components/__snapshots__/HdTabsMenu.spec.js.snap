// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdTabsMenu The component is rendered 1`] = `
<div class="tabs-menu">
  <div class="tabs-menu__wrap">
    <div class="tabs-menu__wrap__inner">
      <div class="flickity-enabled flickity--no-controls" tabindex="0">
        <div class="flickity-viewport" style="height: 0px;">
          <div class="flickity-slider" style="left: 0px; transform: translateX(NaN%);"><button data-index="0" type="button" class="tabs-menu__item btn btn--flat js-tabs-menu-item" style="position: absolute;" aria-hidden="true">
              Jon "Snow"
            </button><button data-index="1" type="button" class="tabs-menu__item btn btn--flat js-tabs-menu-item isActive is-selected" style="position: absolute;">
              Daenerys
            </button><button data-index="2" type="button" class="tabs-menu__item btn btn--flat js-tabs-menu-item" style="position: absolute;" aria-hidden="true">
              The Night King
            </button><button data-index="3" type="button" class="tabs-menu__item btn btn--flat js-tabs-menu-item" style="position: absolute;" aria-hidden="true">
              Arya
            </button><button data-index="4" type="button" class="tabs-menu__item btn btn--flat js-tabs-menu-item" style="position: absolute;" aria-hidden="true">
              Cersei
            </button></div>
        </div><button class="flickity-button flickity-prev-next-button previous" type="button" aria-label="Previous"><svg class="flickity-button-icon" viewBox="0 0 100 100">
            <path d="M 10,50 L 60,100 L 70,90 L 30,50  L 70,10 L 60,0 Z" class="arrow"></path>
          </svg></button><button class="flickity-button flickity-prev-next-button next" type="button" aria-label="Next"><svg class="flickity-button-icon" viewBox="0 0 100 100">
            <path d="M 10,50 L 60,100 L 70,90 L 30,50  L 70,10 L 60,0 Z" class="arrow" transform="translate(100, 100) rotate(180) "></path>
          </svg></button>
      </div>
    </div>
  </div>
</div>
`;
